### ISSUE TITLE
Can IO generics be infered based on its file mode?

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#286]
Can the below type annotations be inferred?

```python
strio: IO[str] = open("foo.json")
binio: IO[bytes] = open("foo.pkl", "rb")
```

Now both of `strio` and `binio` are inferred `IO[AnyStr]`

**Comment by erictraut** [comment]
Thanks for the suggestion. The standard typeshed type stub builtins.pyi doesn't provide a way for a type checker to differentiate. It defines the `open` method as follows.

```
    def open(file: Union[str, bytes, int, _PathLike], mode: str = ..., buffering: int = ..., encoding: Optional[str] = ...,
             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
             opener: Optional[Callable[[str, int], int]] = ...) -> IO[Any]: ...
```

However, it looks like it should be feasible to provide overloads that handle common cases. I've experimented with changing builtins.pyi to the following:
```
    @overload
    def open(file: Union[str, bytes, int, _PathLike], *, buffering: int = ..., encoding: Optional[str] = ...,
             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
             opener: Optional[Callable[[str, int], int]] = ...) -> IO[str]: ... 
    @overload
    def open(file: Union[str, bytes, int, _PathLike], mode: Literal['r', 'w', 'r+', 'a'], buffering: int = ..., encoding: Optional[str] = ...,
             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
             opener: Optional[Callable[[str, int], int]] = ...) -> IO[str]: ... 
    @overload
    def open(file: Union[str, bytes, int, _PathLike], mode: Literal['rb', 'wb', 'r+b', 'ab'], buffering: int = ..., encoding: Optional[str] = ...,
             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
             opener: Optional[Callable[[str, int], int]] = ...) -> IO[bytes]: ... 
    @overload
    def open(file: Union[str, bytes, int, _PathLike], mode: str = ..., buffering: int = ..., encoding: Optional[str] = ...,
             errors: Optional[str] = ..., newline: Optional[str] = ..., closefd: bool = ...,
             opener: Optional[Callable[[str, int], int]] = ...) -> IO[Any]: ...
```

This handles the cases you've mentioned above. I can ship the updated builtins.pyi with the next version of pyright and then see if I can get the maintainers of the typeshed repo to accept the modification into the official builtins.pyi.

**Comment by erictraut** [comment]
This is now fixed in 1.0.73, which I just published.
